contracts/token/ERC721/ERC721.sol

```diff
@@ -433,18 +395,17 @@ contract ERC721 is Context, ERC165, IERC721, IERC721Metadata, IERC721Enumerable
     function _checkOnERC721Received(address from, address to, uint256 tokenId, bytes memory _data)
         private returns (bool)
     {
-        if (!to.isContract()) {
+        if (to.isContract()) {
+            try IERC721Receiver(to).onERC721Received(_msgSender(), from, tokenId, _data) returns (bytes4 retval) {
+                return retval == IERC721Receiver(to).onERC721Received.selector;
+            } catch Error (string memory reason) {
+                revert(reason);
+            } catch {
+                revert("ERC721: transfer to non ERC721Receiver implementer");
```

Comment:
We need something like this to catch every revert reason, including others that would match `Panic(uint256)`, etc.
But the problem with my suggestion below is that I think `revert` wraps it with an `Error(string)` selector. We need to check if this is true, and if so we may need to use assembly `revert`.

```suggestion
            } catch (bytes memory reason) {
                if (reason.length == 0) {
                    revert("ERC721: transfer to non ERC721Receiver implementer");
                } else {
                    revert(reason);
                }
            }
```

Reply:
When the revert calldata contains a "revert reason" string, this is catched by 
```
} catch Error (string memory reason) {
        revert(reason);
}
```
and forwarded correctly.

Catching Panics in a specific way only makes sens if we want to send a specific revert message for them. I'll check that Panic are catched by the genereic `} catch {`, but if yes, the implementation in the PR should be good. It does pass all tests we have, including the ones where the receiver reverts with a message

Reply:
the question is, what error message to we expect to get when:
- the receiver panics
- the receiver reverts without a reason

Reply:
I'm not sure we have any way to distinguish "the receiver reverts without a reason" and "the receiver does not implement `onERC721Received`.

Question is then: what to do when the receiver panics

Reply:
FYI, with master implementation, if the receiver panics, we end up with a revert that matched `expectRevert.unspecified`. Not sure if that contains the panic code, but I personnaly believe that even if it does, it is less expressive then an error that says the erc721receiver caused a panic